Chore: update mcd-seth tutorial (#98)
makerdao/developerguides: mcd/mcd-seth/mcd-seth.md

# Drawing Dai from the Goerli Maker Protocol deployment using Seth

**This guide works under the Goerli [1.11.0 Release](https://changelog.makerdao.com/releases/goerli/1.11.0/index.html) of the system.**

> **You can also use the Rinkey, Ropsten and Goerli deployments in this guide. Just make sure to change the contract addresses from the specific network.**

Reviewer: ```suggestion
**You can also use the Rinkey, Ropsten and Kovan deployments in this guide. Just make sure to change the contract addresses from the specific network.**
```

> **You can also use the Rinkey, Ropsten and Goerli deployments in this guide. Just make sure to change the contract addresses from the specific network.**

Reviewer: ```suggestion
**You can also use the Rinkey, Ropsten and Kovan deployments in this guide. Just make sure to change the contract addresses from the specific network.**
```

This tutorial will cover how to use the tool `seth` to deposit WBTC tokens to draw DAI from the Goerli deployment of MCD as an example, since the process is the same for any other ERC-20 token. You can use the same methodology for any supported token in MCD, by changing the contract addresses to the specific token you want to use.

> <!-- vim-markdown-toc GFM -->

Reviewer: It seems that some vim artifacts were introduced alongside the changes. I would suggest to cleanup those.
```suggestion
```

* [Prerequisites](#prerequisites)
    * [Setting up Seth](#setting-up-seth)
    * [Other tools](#other-tools)
* [Getting tokens](#getting-tokens)
* [Saving contract addresses](#saving-contract-addresses)
* [Token approval](#token-approval)
* [Finally interacting with the Maker Protocol contracts](#finally-interacting-with-the-maker-protocol-contracts)
* [Paying back DAI debt to release collateral](#paying-back-dai-debt-to-release-collateral)

> <!-- vim-markdown-toc -->

Reviewer: Same as above.
```suggestion
```

- `wad`: token unit amount
- `gem`: collateral token adapter
- `ilk`: Vault type
- `urn`: Vault record – keeps track of a Vault
- `ink`: rate \* wad represented in collateral
- `dink`: delta ink – a signed difference value to the current value
- `art`: rate \* wad represented in DAI
- `dart`: delta art – a signed difference value to the current value
- `lad`: Vault owner
- `rat`: collateralization ratio.

After giving permission to the WBTC adapter of MCD to take some of our tokens, it’s time to finally start using the MCD contracts.

We'll be using the [CDP Manager](https://github.com/makerdao/dss-cdp-manager) as the preferred interface to interact with MCD contracts.

We begin by opening an empty Vault, so we can use it to lock collateral into. For this we need to define the type of collateral (WBTC-A) we want to lock in this Vault:
```bash

> export ilk=$(seth --to-bytes32 $(seth --from-ascii 'WBTC-A'))

Reviewer: I'd keep the double quotes here.
```suggestion
export ilk=$(seth --to-bytes32 $(seth --from-ascii "WBTC-A"))
```

Reviewer: see: https://collateral.makerdao.com/smart-contracts-domain/untitled

```bash
seth send $MCD_JOIN_WBTC_A 'join(address, uint256)' $urn $amt
```

**ℹ️ NOTICE:** From this point on, the [join-5](https://goerli.etherscan.io/address/0x3cbE712a12e651eEAF430472c0C1BF1a2a18939D#code) adapter already took care of the fact that WBTC has only 8 decimals, so we can proceed with `wad` normally.

We can check the results with the contract function: `gem(bytes32 ilk, address urn)(uint256)` with:

```bash
seth call $MCD_VAT 'gem(bytes32, address)(uint256)' $ilk $urn | seth --from-wei
```

The output should look like this:

`5.000000000000000000`

An optional, but recommended step is to invoke `jug.drip(ilk)` to make we are not paying undue stability fees. For more detail, please read the guide [Intro to the Rate mechanism](../intro-rate-mechanism/intro-rate-mechanism.md).

```bash
seth send $MCD_JUG 'drip(bytes32)' $ilk
```

The next step is adding the collateral into an urn. This is done through the `CDP Manager` contract.

The function is called `frob(uint256, uint256, uint256)`, which receives the following parameters:

- `uint256 cdp`: the `cdpId`

> - `int256 dink`: delta ink (collateral)
> - `int256 dart`: delta art (Dai).

Reviewer: I'd suggest the following as alternative to the above.
```suggestion
- `int256 dink`: delta ink (collateral) `[wad]`
- `int256 dart`: delta art (Dai) `[wad]`
```

Author: I think it's better if we do both to add some context.

If the `frob` operation is successful, it will adjust the corresponding data in the protected `vat` module. When adding collateral to an `urn`, `dink` needs to be the (positive) amount we want to add and `dart` needs to be the (positive) amount of DAI we want to draw. 

Let’s add our 5 WBTC to the urn, and draw 15000 DAI ensuring that the position is overcollateralized.
We already set up `cdp` before, so we only need to set up `dink` (WBTC deposit) and `dart` (DAI to be drawn):

**ℹ️ NOTICE:** The `vat` uses an internal `dai` representation called “normalized art” that is useful to calculate accrued stability fees. To convert the Dai amount to normalized art, we have to divide it by the current ilk `rate`.

Inside the `vat`, different parameters have different decimal precisions:

> - `dai`: 45 decimals (`rad`)
> - `rate`: 27 decimals (`ray`)
> - `dink`/`dart`: 18 decimals (`wad`).

Reviewer: ```suggestion
- `dink`: 18 decimals (`wad`)
- `dart`: 18 decimals (`wad`)
- `rate`: 27 decimals (`ray`)
- `dai`: 45 decimals (`rad`)
```

Author: Not sure if this should be sorted by the number of decimals. I made it match the usage order in the article.

```bash
export ilk=$(seth --to-bytes32 $(seth --from-ascii "WBTC-A"))
```

Now let’s open the Vault:
```bash
seth send $CDP_MANAGER 'open(bytes32, address)' $ilk $ETH_FROM
```

We need the `cdpId` and `urn` address of our open Vault, so we can interact with the system.

```bash
export cdpId=$(seth call $CDP_MANAGER 'last(address)(uint256)' $ETH_FROM)
export urn=$(seth call $CDP_MANAGER 'urns(uint256)(address)' $cdpId)
```

After acquiring `cdpId` and `urn` address, we can move to the next step. Locking our tokens into the system.

First we are going to make a transaction to the WBTC adapter to actually take 5 of our tokens with the join contract function.
The contract function looks like the following: `join(address urn, uint256 amt)`

- The first parameter is the `urn`, our vault address.
- The second parameter is the token amount in `wad`.

For the sake of readability, we up the `amt` parameter representing the amount of collateral:

```bash
export amt=$(seth --from-fix $WBTC_DECIMALS 5)
```

Then use the following command to use the join function, thus taking 5 WBTC from our account and sending to `urn` address.

```bash

> seth send $MCD_JOIN_WBTC_A "join(address, uint256)" $urn $amt

Reviewer: ```suggestion
seth send $MCD_JOIN_WBTC_A 'join(address, uint256)' $urn $amt
```

> seth send $MCD_JOIN_WBTC_A "join(address, uint256)" $urn $amt

Reviewer: ```suggestion
seth send $MCD_JOIN_WBTC_A 'join(address, uint256)' $urn $amt
```

- `uint256 cdp` &ndash; the `cdpId`
- `int256 dink` &ndash; delta ink (collateral)
- `int256 dart` &ndash; delta art (Dai).

If the `frob` operation is successful, it will adjust the corresponding data in the protected `vat` module. When adding collateral to an `urn`, `dink` needs to be the (positive) amount we want to add and `dart` needs to be the (positive) amount of DAI we want to draw. 

Let’s add our 5 WBTC to the urn, and draw 15000 DAI ensuring that the position is overcollateralized.
We already set up `cdp` before, so we only need to set up `dink` (WBTC deposit) and `dart` (DAI to be drawn):

**NOTICE:** The `vat` uses an internal `dai` representation called &ldquo;normalized art&rdquo; that is useful to calculate accrued stability fees. To convert the Dai amount to normalized art, we have to divide it by the current ilk `rate`.

Inside the `vat`, different parameters have different decimal precisions:

- `dai`: 45 decimals (`rad`)
- `rate`: 27 decimals (`ray`)
- `dink`/`dart`: 18 decimals (`wad`)

Learn more about naming in MCD [here](https://github.com/makerdao/dss/wiki/Glossary#general).

```bash

> export RAD_DECIMALS=45
> export RAY_DECIMALS=27
> export WAD_DECIMALS=18

Reviewer: ```suggestion
export WAD_DECIMALS=18
export RAY_DECIMALS=27
export RAD_DECIMALS=45
```

We now allow the Dai adapter to move Dai from VAT to our address:
```bash
seth send $MCD_VAT 'hope(address)' $MCD_JOIN_DAI
```

And finally we exit the internal `dai` to the ERC-20 DAI:
```bash
seth send $MCD_JOIN_DAI "exit(address, uint256)" $ETH_FROM $(seth --to-wei 15000 eth)
```

And to check the DAI balance of our account:
```bash
seth call $MCD_DAI 'balanceOf(address)(uint256)' $ETH_FROM | seth --from-wei
```

Expected output:

`15000.000000000000000000`

If everything checks out, congratulations: you have just acquired some multi-collateral DAI on Goerli!

## Paying back DAI debt to release collateral

To pay back your DAI and release the locked collateral, follow the following steps. 

**IMPORTANT:** Please make sure to **obtain some additional Dai** (from another account or from another vault) because chances are interest will have accumulated in the meantime.

To force stability fee accumulation, anyone can invoke `jug.drip(ilk)`:

```bash
seth send $MCD_JUG 'drip(bytes32)' $ilk
```

First thing is to determine what is our debt, including the accrued stability fee:

```bash

> export RAD_DECIMALS=45
> export RAY_DECIMALS=27
> export WAD_DECIMALS=18

Reviewer: Same as above.
```suggestion
export WAD_DECIMALS=18
export RAY_DECIMALS=27
export RAD_DECIMALS=45
```

export art=$(seth call $MCD_VAT 'urns(bytes32, address)(uint256 ink, uint256 art)' $ilk $urn | \
    sed -n 2p | seth --from-wei)
export rate=$(seth call $MCD_VAT \
    'ilks(bytes32)(uint256 Art, uint256 rate, uint256 spot, uint256 line, uint256 dust)' $ilk | \
    sed -n 2p | seth --to-fix $RAY_DECIMALS)
export debt=$(bc<<<"$art*$rate")
export debtWadRound=$(bc<<<"($art*$rate*10^${WAD_DECIMALS})/1 + 1")
```

- `art`: internal vault debt representation
- `rate`: accumulated stability fee from the system
- `debt`: vault debt in Dai
- `debtWadRound`: vault debt in wad (i.e. multiplied by 10^18), added by 1 wad to avoid rounding issues.

Then we need to approve the transfer of DAI tokens to the adapter. Call the `approve` function of the DAI ERC-20 token contract with the right parameters:

```bash
seth send $MCD_DAI 'approve(address, uint256)' $MCD_JOIN_DAI $debtWadRound
```

If we want to be sure that our approval transaction succeeded, we can check the results with this command:

```bash
seth call $MCD_DAI 'allowance(address, address)(uint256)' $ETH_FROM $MCD_JOIN_DAI | seth --from-wei
```

Output:

`15000.041850037339693452`

Now to actually join the Dai to the adapter:

```bash

> seth send $MCD_JOIN_DAI "join(address, uint256)" $urn $debtWadRound

Reviewer: ```suggestion
seth send $MCD_JOIN_DAI 'join(address, uint256)' $urn $debtWadRound
```
```suggestion
seth send $MCD_JOIN_DAI "join(address, uint256)" $urn $debtWadRound
```

> seth send $MCD_JOIN_DAI "join(address, uint256)" $urn $debtWadRound

Reviewer: ```suggestion
seth send $MCD_JOIN_DAI 'join(address, uint256)' $urn $debtWadRound
```
```suggestion
seth send $MCD_JOIN_DAI "join(address, uint256)" $urn $debtWadRound
```

```

To make sure it all worked:

```bash
seth call $MCD_VAT 'dai(address)(uint256)' $urn | seth --to-fix $RAD_DECIMALS
```

Output:

`15000.041850037339693452000000000000000000000000000`

Now, onto actually getting our collateral back. `dart` and `dink`, as the `d` in their abbreviation stands for delta, are inputs for changing a value, and thus they can be negative. When we want to lower the amount of DAI drawn from the `urn`, we lower the art parameter of the `urn`.

We only need to set up the `dink` and `dart` variables.
Using [mcd-cli](https://github.com/makerdao/mcd-cli#installation), we create our two negative `$dink` and `$dart`:

```bash

> dink=$(seth --to-uint256 $(mcd --to-hex $(seth --to-wei -5 eth)))

Reviewer: I'd suggest to consider the following change as It seems that at the time of writing of this tutorial `seth` couldn't manage negative integers, that now seem to be supported.
The following should return the same output:
```suggestion
dink=$(seth --to-int256 $(seth --to-wei -5 eth))
```

> dart=$(seth --to-uint256 $(mcd --to-hex -$(seth --to-wei $art eth)))

Reviewer: Same as above.
```suggestion
dart=$(seth --to-int256 -$(seth --to-wei $art eth))
```

Author: Does this mean we can we get rid of the `mcd-cli` dependency?

Reviewer: yep

```

Again, we need to use the `frob` operation to change these parameters: `frob(uint256 cdpId, address from, int dink, int dart)`

```bash
seth send $CDP_MANAGER "frob(uint256, int256, int256)" $cdpId $dink $dart
```

This doesn’t mean we have already got back your tokens yet. Our account’s WBTC balance is not yet back to the original amount:

```bash
seth call $WBTC 'balanceOf(address)(uint256)' $ETH_FROM | seth --to-fix $WBTC_DECIMALS
```

Output:

`0.00000000`

The WBTC is still assigned to the Vault, so we need to move them to our address:
```bash

> export wadC=$(seth --to-wei 5 eth)

Reviewer: I'd suggest to rename it `wad`.
```suggestion
export wad=$(seth --to-wei 5 eth)
```

> seth send $CDP_MANAGER 'flux(uint256, address, uint256)' $cdpId $ETH_FROM $wadC

Reviewer: Same as above.
```suggestion
seth send $CDP_MANAGER 'flux(uint256, address, uint256)' $cdpId $ETH_FROM $wad
```

```

**NOTICE:** We are about to interact with the [join-5](https://goerli.etherscan.io/address/0x3cbE712a12e651eEAF430472c0C1BF1a2a18939D#code) adapter once again, so we need to bring `$WBTC_DECIMALS` back into the equation.

From there exit the WBTC adapter to get back our tokens:

```bash
export WBTC_DECIMALS=8

export amt=$(seth --from-fix $WBTC_DECIMALS 5)

> seth send $MCD_JOIN_WBTC_A "exit(address, uint256)" $ETH_FROM $amt

Reviewer: ```suggestion
seth send $MCD_JOIN_WBTC_A 'exit(address, uint256)' $ETH_FROM $amt
```

> seth send $MCD_JOIN_WBTC_A "exit(address, uint256)" $ETH_FROM $amt

Reviewer: ```suggestion
seth send $MCD_JOIN_WBTC_A 'exit(address, uint256)' $ETH_FROM $amt
```

- `lad`: Vault owner
- `rat`: collateralization ratio

After giving permission to the WBTC adapter of MCD to take some of our tokens, it’s time to finally start using the MCD contracts.

We'll be using the [CDP Manager](https://github.com/makerdao/dss-cdp-manager) as the preferred interface to interact with MCD contracts.

We begin by opening an empty Vault, so we can use it to lock collateral into. For this we need to define the type of collateral (WBTC-A) we want to lock in this Vault:
```bash
export ilk=$(seth --to-bytes32 $(seth --from-ascii "WBTC-A"))
```

Now let’s open the Vault:
```bash
seth send $CDP_MANAGER 'open(bytes32, address)' $ilk $ETH_FROM
```

We need the `cdpId` and `urn` address of our open Vault, so we can interact with the system.

```bash
export cdpId=$(seth call $CDP_MANAGER 'last(address)(uint256)' $ETH_FROM)
export urn=$(seth call $CDP_MANAGER 'urns(uint256)(address)' $cdpId)
```

After acquiring `cdpId` and `urn` address, we can move to the next step. Locking our tokens into the system.

First we are going to make a transaction to the WBTC adapter to actually take 5 of our tokens with the join contract function.

> The contract function looks like the following: `join(address urn, uint256 wad)`

Reviewer: I'd suggest to maybe add a note to specify that the WBTC adapter uses a non standard GemJoin ([GemJoin5](https://github.com/makerdao/dss-gem-joins/blob/master/src/join-5.sol)).
```suggestion
The contract function looks like the following: `join(address urn, uint256 amt)`
```

- The first parameter is the `urn`, our vault address.

> - The second parameter is the token amount in `wad`.

Reviewer: ```suggestion
- The second parameter is the token amount `amt`.
```

For the sake of readability, we up the `amt` parameter representing the amount of collateral:

```bash
export amt=$(seth --from-fix $WBTC_DECIMALS 5)
```

> Then use the following command to use the join function, thus taking 150 WBTC from our account and sending to `urn` address.

Reviewer: ```suggestion
Then use the following command to use the join function, thus taking 5 WBTC from our account and sending to `urn` address.
```

```bash
seth send $MCD_JOIN_WBTC_A "join(address, uint256)" $urn $amt
```

> **NOTICE:** From this point on, the join adapter already took care of the fact that WBTC has only 8 decimals, so we can proceed with `wad` normally.

Reviewer: ```suggestion
**NOTICE:** From this point on, the join-5 adapter already took care of the fact that WBTC has only 8 decimals, so we can proceed with `wad` normally.
```

We can check the results with the contract function: `gem(bytes32 ilk, address urn)(uint256)` with:

```bash
seth call $MCD_VAT 'gem(bytes32, address)(uint256)' $ilk $urn | seth --from-wei
```

The output should look like this:

`5.000000000000000000`

An optional, but recommended step is to invoke `jug.drip(ilk)` to make we are not paying undue stability fees. For more detail, please read the guide [Intro to the Rate mechanism](../intro-rate-mechanism/intro-rate-mechanism.md).

```bash
seth send $MCD_JUG 'drip(bytes32)' $ilk
```

The next step is adding the collateral into an urn. This is done through the `CDP Manager` contract.

The function is called `frob(uint256, uint256, uint256)`, which receives the following parameters:

- `uint256 cdp` - the `cdpId`
- `int256 dink` - delta ink (collateral)
- `int256 dart` - delta art (Dai).

If the `frob` operation is successful, it will adjust the corresponding data in the protected `vat` module. When adding collateral to an `urn`, `dink` needs to be the (positive) amount we want to add and `dart` needs to be the (positive) amount of DAI we want to draw. 

> Let’s add our 150 WBTC to the urn, and draw 15000 DAI ensuring that the position is overcollateralized.

Reviewer: ```suggestion
Let’s add our 5 WBTC to the urn, and draw 15000 DAI ensuring that the position is overcollateralized.
```

We already set up `cdp` before, so we only need to set up `dink` (WBTC deposit) and `dart` (DAI to be drawn):

```bash

> dink=$(seth --to-wei 5 eth)
> rate=$(seth call $MCD_VAT 'ilks(bytes32)(uint256 Art, uint256 rate, uint256 spot, uint256 line, uint256 dust)' $ilk | \
>     sed -n 2p | seth --to-fix 27)
> dart=$(bc<<<"scale=18; art=(15000/$rate*10^18+1); scale=0; art/1")

Reviewer: ```suggestion
export dink=$(seth --to-wei 5 eth)
export rate=$(seth call $MCD_VAT 'ilks(bytes32)(uint256 Art, uint256 rate, uint256 spot, uint256 line, uint256 dust)' $ilk | \
    sed -n 2p | seth --to-fix 27)
export dart=$(bc<<<"scale=18; art=(15000/$rate*10^18+1); scale=0; art/1")
```
